src/Formula/reaction.ts

```diff
@@ -1,97 +1,114 @@
-  },
-  hydro: {
-    electrocharged: trans.electrocharged,
-    shattered: trans.shattered,
-    bloom: trans.bloom,
-  },
-  pyro: {
-    overloaded: trans.overloaded,
-    shattered: trans.shattered,
-    burning: trans.burning,
-    burgeon: trans.burgeon,
-  },
-  cryo: {
-    superconduct: trans.superconduct,
-    shattered: trans.shattered,
-  },
-  dendro: {
-    burning: trans.burning,
-    bloom: trans.bloom,
+export function getReactions(charElement: ElementKey) {
+  return {
+    electroSwirl: infoMut(
+      checkCharEleAndInfusion(charElement, ["anemo"], trans.swirl.electro)
+      , { key: "electro_swirl_hit" }
+    ),
+    pyroSwirl: infoMut(
+      checkCharEleAndInfusion(charElement, ["anemo"], trans.swirl.pyro),
+      { key: "pyro_swirl_hit" }
+    ),
+    cryoSwirl: infoMut(
+      checkCharEleAndInfusion(charElement, ["anemo"], trans.swirl.cryo),
+      { key: "cryo_swirl_hit" }
+    ),
+    hydroSwirl: infoMut(
+      checkCharEleAndInfusion(charElement, ["anemo"], trans.swirl.hydro),
+      { key: "hydro_swirl_hit" }
+    ),
+    overloaded: infoMut(
+      checkCharEleAndInfusion(charElement, ["pyro", "electro", "anemo"], trans.overloaded),
+      { key: "overloaded_hit" }
+    ),
+    electrocharged: infoMut(
+      checkCharEleAndInfusion(charElement, ["hydro", "electro", "anemo"], trans.electrocharged),
+      { key: "electrocharged_hit" }
+    ),
+    superconduct: infoMut(
+      checkCharEleAndInfusion(charElement, ["cryo", "electro", "anemo"], trans.superconduct),
+      { key: "superconduct_hit" }
+    ),
+    // Technically shatter does not apply to all characters, but the logic would be too complicated for no reason
+    shattered: infoMut(trans.shattered, { key: "shattered_hit" }),
+    burning: infoMut(
+      checkCharEleAndInfusion(charElement, ["pyro", "dendro", "anemo"], trans.burning),
+      { key: "burning_hit" }
+    ),
+    bloom: infoMut(
+      checkCharEleAndInfusion(charElement, ["hydro", "dendro", "anemo"], trans.bloom),
+      { key: "bloom_hit" }
+    ),
+    burgeon: infoMut(
+      checkCharEleAndInfusion(charElement, ["pyro", "anemo"], trans.burgeon),
+      { key: "burgeon_hit" }
+    ),
+    hyperbloom: infoMut(
+      checkCharEleAndInfusion(charElement, ["electro", "anemo"], trans.hyperbloom),
+      { key: "hyperbloom_hit" }
+    ),
+    crystallize: infoMut(
+      checkCharEleAndInfusion(charElement, ["geo"], crystallizeHit), { key: "crystallize" }),
+    ...Object.fromEntries(absorbableEle.map(e => [
+      `${e}Crystallize`,
+      infoMut(
+        checkCharEleAndInfusion(charElement, ["geo", e], prod(percent(2.5), crystallizeHit)),
+        { key: `${e}_crystallize`, variant: e })
+    ])),
   }
 }
+
+// Check the character's element and infusion against an array of elements. If any matches, the nodes value will be `nodeToReturn`.
+function checkCharEleAndInfusion(charElement: ElementKey, elementsToMatch: ElementKey[], nodeToReturn: NumNode): NumNode {
```

Comment:
Why are we passing in character's element key here? why wouldnt you be able to access `input.charEle`?